File: src/main/java/org/hisp/dhis/rules/RuleEngine.java
package org.hisp.dhis.rules;

import org.hisp.dhis.lib.expression.Expression;
import org.hisp.dhis.lib.expression.spi.IllegalExpressionException;
import org.hisp.dhis.lib.expression.spi.ParseException;
import org.hisp.dhis.lib.expression.spi.ValueType;
import org.hisp.dhis.rules.models.Rule;
import org.hisp.dhis.rules.models.RuleEffect;
import org.hisp.dhis.rules.models.RuleEffects;
import org.hisp.dhis.rules.models.RuleEngineValidationException;
import org.hisp.dhis.rules.models.RuleEnrollment;
import org.hisp.dhis.rules.models.RuleEvent;
import org.hisp.dhis.rules.models.RuleValidationResult;
import org.hisp.dhis.rules.models.TriggerEnvironment;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;

// ToDo: logging
public final class RuleEngine
{
    @Nonnull
    private final RuleEngineContext ruleEngineContext;

    @Nonnull
    private final List<RuleEvent> ruleEvents;

    @Nullable
    private final RuleEnrollment ruleEnrollment;

    @Nullable
    private TriggerEnvironment triggerEnvironment;

    RuleEngine( @Nonnull RuleEngineContext ruleEngineContext,
        @Nonnull List<RuleEvent> ruleEvents,
        @Nullable RuleEnrollment ruleEnrollment, @Nullable TriggerEnvironment triggerEnvironment )
    {
        this.ruleEngineContext = ruleEngineContext;
        this.ruleEvents = ruleEvents;
        this.ruleEnrollment = ruleEnrollment;
        this.triggerEnvironment = triggerEnvironment;
    }

    @Nonnull
    public List<RuleEvent> events()
    {
        return ruleEvents;
    }

    @Nullable
    public RuleEnrollment enrollment()
    {
        return ruleEnrollment;
    }

    @Nullable
    public TriggerEnvironment triggerEnvironment()
    {
        return triggerEnvironment;
    }

    @Nonnull
    public RuleEngineContext executionContext()
    {
        return ruleEngineContext;
    }

    @Nonnull
    public Callable<List<RuleEffect>> evaluate( @Nonnull RuleEvent ruleEvent )
    {
        return evaluate( ruleEvent, ruleEngineContext.rules() );
    }

    @Nonnull
    public Callable<List<RuleEffect>> evaluate( @Nonnull RuleEvent ruleEvent, @Nonnull List<Rule> rulesToEvaluate )
    {
        if ( ruleEvent == null )
        {
            throw new IllegalArgumentException( "ruleEvent == null" );
        }

        Map<String, RuleVariableValue> valueMap = RuleVariableValueMapBuilder.target( ruleEvent )
            .ruleVariables( ruleEngineContext.ruleVariables() )
            .ruleEnrollment( ruleEnrollment )
            .triggerEnvironment( triggerEnvironment )
            .ruleEvents( ruleEvents )
            .constantValueMap( ruleEngineContext.constantsValues() )
            .build();

        return new RuleEngineExecution( ruleEvent, rulesToEvaluate, valueMap, ruleEngineContext.supplementaryData() );
    }

    @Nonnull
    public Callable<List<RuleEffect>> evaluate( @Nonnull RuleEnrollment ruleEnrollment,
        @Nonnull List<Rule> rulesToEvaluate )
    {
        if ( ruleEnrollment == null )
        {
            throw new IllegalArgumentException( "ruleEnrollment == null" );
        }

        Map<String, RuleVariableValue> valueMap = RuleVariableValueMapBuilder.target( ruleEnrollment )
            .ruleVariables( ruleEngineContext.ruleVariables() )
            .triggerEnvironment( triggerEnvironment )
            .ruleEvents( ruleEvents )
            .constantValueMap( ruleEngineContext.constantsValues() )
            .build();

        return new RuleEngineExecution( ruleEnrollment, rulesToEvaluate, valueMap, ruleEngineContext.supplementaryData() );
    }

    @Nonnull
    public Callable<List<RuleEffects>> evaluate()
    {
        RuleVariableValueMap valueMap = RuleVariableValueMapBuilder.target()
            .ruleVariables( ruleEngineContext.ruleVariables() )
            .ruleEnrollment( ruleEnrollment )
            .triggerEnvironment( triggerEnvironment )
            .ruleEvents( ruleEvents )
            .constantValueMap( ruleEngineContext.constantsValues() )
            .multipleBuild();

        return new RuleEngineMultipleExecution( ruleEngineContext.rules(), valueMap,
            ruleEngineContext.supplementaryData() );
    }

    @Nonnull
    public Callable<List<RuleEffect>> evaluate( @Nonnull RuleEnrollment ruleEnrollment )
    {
        return evaluate( ruleEnrollment, ruleEngineContext.rules() );
    }

    @Nonnull
    public RuleValidationResult evaluate( String expression )
    {
        // Rule condition expression should be evaluated against Boolean
        return getExpressionDescription(expression, Expression.Mode.RULE_ENGINE_CONDITION);
    }

    @Nonnull
    public RuleValidationResult evaluateDataFieldExpression( String expression )
    {
        // Rule action data field field should be evaluated against all i.e Boolean, String, Date and Numerical value
        return getExpressionDescription( expression, Expression.Mode.RULE_ENGINE_ACTION);
    }

    private RuleValidationResult getExpressionDescription(String expression, Expression.Mode mode)
    {
        try {
            Map<String, ValueType> validationMap = new HashMap<>();
            for (Map.Entry<String, DataItem> e : ruleEngineContext.getDataItemStore().entrySet()) {
                validationMap.put(e.getKey(), e.getValue().getValueType().toValueType());
            }
            new Expression(expression, mode).validate( validationMap );

            Map<String, String> displayNames = new HashMap<>();
            for (Map.Entry<String, DataItem> e : ruleEngineContext.getDataItemStore().entrySet()) {
                displayNames.put(e.getKey(), e.getValue().getDisplayName());
            }
            String description = new Expression(expression, mode).describe(displayNames);
            return RuleValidationResult.builder().isValid( true ).description(description).build();
        } catch (IllegalExpressionException | ParseException ex) {
            return RuleValidationResult.builder()
                    .isValid(false)
                    .exception(new RuleEngineValidationException(ex.getMessage()))
                    .errorMessage(ex.getMessage())
                    .build();
        }
    }

    public static class Builder
    {
        @Nonnull
        private final RuleEngineContext ruleEngineContext;

        @Nullable
        private List<RuleEvent> ruleEvents;

        @Nullable
        private RuleEnrollment ruleEnrollment;

        @Nullable
        private TriggerEnvironment triggerEnvironment;

        Builder( @Nonnull RuleEngineContext ruleEngineContext )
        {
            this.ruleEngineContext = ruleEngineContext;
        }

        @Nonnull
        public Builder events( @Nonnull List<RuleEvent> ruleEvents )
        {
            if ( ruleEvents == null )
            {
                throw new IllegalArgumentException( "ruleEvents == null" );
            }

            this.ruleEvents = Collections.unmodifiableList( new ArrayList<>( ruleEvents ) );
            return this;
        }

        @Nonnull
        public Builder enrollment( @Nonnull RuleEnrollment ruleEnrollment )
        {
            if ( ruleEnrollment == null )
            {
                throw new IllegalArgumentException( "ruleEnrollment == null" );
            }

            this.ruleEnrollment = ruleEnrollment;
            return this;
        }

        @Nonnull
        public Builder triggerEnvironment( @Nonnull TriggerEnvironment triggerEnvironment )
        {
            if ( triggerEnvironment == null )
            {
                throw new IllegalArgumentException( "triggerEnvironment == null" );
            }

            this.triggerEnvironment = triggerEnvironment;
            return this;
        }

        @Nonnull
        public RuleEngine build()
        {
            if ( ruleEvents == null )
            {
                ruleEvents = Collections.unmodifiableList( new ArrayList<RuleEvent>() );
            }

            return new RuleEngine( ruleEngineContext, ruleEvents, ruleEnrollment, triggerEnvironment );
        }
    }
}

File: src/main/java/org/hisp/dhis/rules/RuleEngine.kt
package org.hisp.dhis.rules

import org.hisp.dhis.lib.expression.Expression
import org.hisp.dhis.lib.expression.spi.IllegalExpressionException
import org.hisp.dhis.lib.expression.spi.ParseException
import org.hisp.dhis.lib.expression.spi.ValueType
import org.hisp.dhis.rules.models.*
import java.util.Date
import java.util.concurrent.Callable

data class RuleEngine(
    val executionContext: RuleEngineContext,
    val events: List<RuleEvent> = emptyList(),
    val enrollment: RuleEnrollment? = null,
    val triggerEnvironment: TriggerEnvironment? = TriggerEnvironment.SERVER) {

    @Deprecated("use kotlin property")
    fun executionContext(): RuleEngineContext {
        return executionContext
    }

    @Deprecated("use kotlin property")
    fun events(): List<RuleEvent> {
        return events
    }

    fun evaluate(ruleEvent: RuleEvent): Callable<List<RuleEffect>> {
        return evaluate(ruleEvent, executionContext.rules)
    }

    fun evaluate(ruleEvent: RuleEvent, rulesToEvaluate: List<Rule>): Callable<List<RuleEffect>> {
        val valueMap = RuleVariableValueMapBuilder.target(ruleEvent)
                .ruleVariables(executionContext.ruleVariables)
                .ruleEnrollment(enrollment)
                .triggerEnvironment(triggerEnvironment)
                .ruleEvents(events)
                .constantValueMap(executionContext.constantsValues)
                .build()
        return RuleEngineExecution(ruleEvent, null, rulesToEvaluate, valueMap, executionContext.supplementaryData)
    }

    fun evaluate(ruleEnrollment: RuleEnrollment,
                 rulesToEvaluate: List<Rule>): Callable<List<RuleEffect>> {
        val valueMap = RuleVariableValueMapBuilder.target(ruleEnrollment)
                .ruleVariables(executionContext.ruleVariables)
                .triggerEnvironment(triggerEnvironment)
                .ruleEvents(events)
                .constantValueMap(executionContext.constantsValues)
                .build()
        return RuleEngineExecution(null, ruleEnrollment, rulesToEvaluate, valueMap, executionContext.supplementaryData)
    }

    fun evaluate(): List<RuleEffects> {
        val valueMap = RuleVariableValueMapBuilder.target()
                .ruleVariables(executionContext.ruleVariables)
                .ruleEnrollment(enrollment)
                .triggerEnvironment(triggerEnvironment)
                .ruleEvents(events)
                .constantValueMap(executionContext.constantsValues)
                .multipleBuild()
        return RuleEngineMultipleExecution(executionContext.rules, valueMap,
                executionContext.supplementaryData).call()
    }

    fun evaluate( ruleEnrollment: RuleEnrollment): Callable<List<RuleEffect>> {
        return evaluate(ruleEnrollment, executionContext.rules)
    }

    fun evaluate(expression: String): RuleValidationResult {
        // Rule condition expression should be evaluated against Boolean
        return getExpressionDescription(expression, Expression.Mode.RULE_ENGINE_CONDITION)
    }

    fun evaluateDataFieldExpression(expression: String): RuleValidationResult {
        // Rule action data field field should be evaluated against all i.e Boolean, String, Date and Numerical value
        return getExpressionDescription(expression, Expression.Mode.RULE_ENGINE_ACTION)
    }

    private fun getExpressionDescription(expression: String, mode: Expression.Mode): RuleValidationResult {
        return try {
            val validationMap: MutableMap<String, ValueType> = HashMap()
            val dataItemStore = executionContext.dataItemStore
            for ((key, value) in dataItemStore!!) {
                validationMap[key] = value.valueType.toValueType()
            }
            Expression(expression, mode, false).validate(validationMap)
            val displayNames: MutableMap<String, String> = HashMap()
            for ((key, value) in dataItemStore) {
                displayNames[key] = value.displayName
            }
            val description = Expression(expression, mode, false).describe(displayNames)
            RuleValidationResult( valid=true, description = description)
        } catch (ex: IllegalExpressionException) {
            RuleValidationResult(valid = false, exception = RuleEngineValidationException(ex), errorMessage = ex.message)
        } catch (ex: ParseException) {
            RuleValidationResult(valid = false, exception = RuleEngineValidationException(ex), errorMessage = ex.message)
        }
    }

    @Deprecated("use data copy")
    class Builder internal constructor(private val ruleEngineContext: RuleEngineContext) {
        private var ruleEvents: List<RuleEvent> = emptyList()
        private var ruleEnrollment: RuleEnrollment? = null
        private var triggerEnvironment: TriggerEnvironment? = null
        fun events(ruleEvents: List<RuleEvent>): Builder {
            this.ruleEvents = ruleEvents
            return this
        }

        fun enrollment(ruleEnrollment: RuleEnrollment): Builder {
            this.ruleEnrollment = ruleEnrollment
            return this
        }

        fun triggerEnvironment(triggerEnvironment: TriggerEnvironment): Builder {
            this.triggerEnvironment = triggerEnvironment
            return this
        }

        fun build(): RuleEngine {
            return RuleEngine(ruleEngineContext, ruleEvents, ruleEnrollment, triggerEnvironment)
        }
    }
}
